src/ast/operations.rs

```diff
@@ -10,20 +10,29 @@ use crate::ast::visitor;
 use crate::ast::visitor::Visitor;
 
 /// Extract the names bound to a given __all__ assignment.
-pub fn extract_all_names(stmt: &Stmt, scope: &Scope, bindings: &[Binding]) -> Vec<String> {
-    fn add_to_names(names: &mut Vec<String>, elts: &[Expr]) {
+pub fn extract_all_names(
+    stmt: &Stmt,
+    scope: &Scope,
+    bindings: &[Binding],
+) -> (Vec<String>, bool, bool) {
```

Comment:
My gut reaction is that this operation is fast (and rare) enough that I'd rather split this into three separate functions. What do you think?

I was going to suggest making this `-> Result<Vec<String>, CustomError>`, and having the error states captured by `CustomError`. But it looks like the errors aren't mutually exclusive (both can occur at once), and that even if it "errors", we still want to return the names.

Reply:
`extract_all_names` extracts the names where it can, and ignores other cases (such as when the type of `_all__` is not a list of tuple. Keeping track on if it has encountered any value to ignore is arguably nog out of its scope of complicating its logic (it should not  be tied to PLE directly, this responsibility is in `ast.rs`. 

`Result<Vec<str>, ..>` wouldn't work, as both the return value and this state are used. We could also group the bools into a state struct and return it in a tuple.

Splitting up the functions is also an option. Currently this is the only place where we go through the values in `__all__`. I do not think they would be distinct enough to have code repetiton. On the other hand, the logic for checking might evolve and that should not happen in `operarions.rs` (could be concern for later).

Reply:
Yeah, totally understand. I was suggesting that we split up the function. I hadn't gone through the exercise, so I wasn't sure how much code would be duplicated, but sometimes in those cases, I just prefer to write-it-twice.

Maybe instead of returning two booleans, we return a `bitflags` flag, so that the responsibilities are clearer?

Reply:
Thanks. Refactored using bitflags. The responsibilities are now much clearer, do you agree?

Reply:
Awesome, thank you :)